fedbiomed/researcher/job.py

```diff
@@ -475,22 +494,131 @@ def start_nodes_training_round(self,
+        )
+        # Upload the shared information that all nodes will download.
+        if aux_shared:
+            path = os.path.join(
+                self._keep_files_dir, f"aux_var_shared_{uuid.uuid4()}.mpk"
+            )
+            Serializer.dump(aux_shared, path)
+            url_shared = self.repo.upload_file(path)["file"]
+        else:
+            url_shared = None
+        # Upload the node-specific information, with node-specific urls.
+        url_bynode = {}  # type: Dict[uuid.UUID, str]
+        for node_id, node_aux in aux_bynode.items():
+            if not node_aux:
+                continue
+            path = os.path.join(
+                self._keep_files_dir,
+                f"aux_var_node_{node_id}_{uuid.uuid4()}.mpk"
+            )
+            Serializer.dump(node_aux, path)
+            url_bynode[node_id] = self.repo.upload_file(path)["file"]
+        # Return the urls of the uploaded files.
+        return url_shared, url_bynode
+
+    @staticmethod
+    def _prepare_agg_optimizer_aux_var(
+        aux_var: Dict[str, Dict[str, Any]],
+        nodes: List[uuid.UUID],
+    ) -> Tuple[
+        Dict[str, Dict[str, Any]],
+        Dict[uuid.UUID, Dict[str, Dict[str, Any]]],
+    ]:
+        """Collect and structure researcher-side Optimizer auxiliary variables.
+
+        Args:
+            aux_var: Auxiliary variables with to structure into multiple dicts,
+                from `{mod_name: (shared_dict | {node_id: node_dict})}` to
+                `{mod_name: shared_dict}` & `{node_id: {mod_name: node_dict}}`.
+            nodes: Ids of the nodes to whom auxiliary variables should be
+                sent. This is used to drop information of non-participating
+                nodes.
+
+        Returns:
+            aux_shared: Dict containing auxiliary variables that are shared
+                across all nodes, with `{mod_name: shared_dict}` format.
+            aux_bynode: Dict containing node-wise dicts of node-specific
+                auxiliary variables, with `{node_id: {mod_name: node_dict}}`
+                format.
+        """
+        aux_shared = {}  # type: Dict[str, Dict[str, Any]]
+        aux_bynode = {}  # type: Dict[uuid.UUID, Dict[str, Dict[str, Any]]]
+        # Iterate over nodes and plug-in-module-wise auxiliary variables.
+        for node_id in nodes:
+            aux_bynode[node_id] = {}
+            for mod_name, mod_info in aux_var.items():
+                # Case of node-specfic information.
+                if node_aux := mod_info.get(str(node_id)):
+                    aux_bynode[node_id][mod_name] = node_aux
+                # Case of global information shared with all nodes.
+                elif mod_name not in aux_shared:
+                    aux_shared[mod_name] = mod_info
+        # Return the restructured auxiliary variables dicts.
+        return aux_shared, aux_bynode
+
+    def get_received_optimizer_aux_var_from_round(
+        self,
```

Comment:
I really like what you did: I am just a bit concerned by the name of the function: I like the name `get_received_optimizer_aux_var_from_round`, but should we really have to  specify `_from_round` ?

Reply:
To me, `process` implies some side effects, while here this _is_ a getter, with some minor-overhead restructuring but no side effects. Hence my choosing `get`. That being said, I would be okay with something in the vein of `access` or `extract`, which might emphasize that this has some (limited) cost.

Reply:
I think `extract` is a good alternative: let s call it : `extract_received_optimizer_aux_var_from_round` !